### PR TITLE
fixes #11191 - add setting for customized homepage to redirect to after login

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by domcleal** [comment]
This should not be a global setting, as it would only exacerbate issues such as #1632.  This also isn't extensible sensibly via plugins as suggested in the ticket description..

**Comment by isratrade** [comment]
@domcleal, the fusor or RHCI plugin needs to modify this value to be `r/#/deployments/new/start`

I think you wrote the wrong issue??

**Comment by domcleal** [comment]
Sorry, it linked the PR.  I mean ticket 1632: http://projects.theforeman.org/issues/1632

This will only work for one plugin though, which is why we have the customisable dashboard.  If you're writing some plugin to change Foreman's behaviour, I think it's better just to patch it rather than have magical settings that are only good for one plugin...

**Comment by isratrade** [comment]
@domcleal, I consider this PR a feature and not a magical setting for one plugin. Yes, the impetus is that I can overwrite the root path to be `r/#/deployments/new/start` but I think other users may want to be able to redirect to a specified page such as `hosts/new` or only other valid route.
Since this PR **doesn't change any default behavior**, I don't understand the opposition, just because a non-admin user may not have permission to view. It's the same case as today (as the issue 1632 mentions) when it redirects to `hosts_path`, so I don't see any degradation. Furthermore, if other plugins want to override the homepage, I think it's good for Foreman to provide a simple option, rather than expecting each plugin developer to override `dashboard#index` or some other hack

**Comment by ares** [comment]
I agree with @domcleal, if plugin is meant to set a global value, how can two such plugins live together? How about keeping this global setting that admin can change and add one more setting per user so every user can override?

**Comment by isratrade** [comment]
@ares, @domcleal, I modified the PR to make it a user setting (attribute) rather than a global setting.

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -125,6 +125,8 @@ def login_user(user)
     session[:original_uri] = nil
     set_current_taxonomies(user, {:session => session})
     TopbarSweeper.expire_cache(self)
-    redirect_to (uri || hosts_path)
+    homepage = User.current.try(:homepage)
+    homepage = "/#{homepage}" if homepage && !homepage.starts_with?('/')
```
Comment (domcleal) [inline app/controllers/users_controller.rb:129]: This code's duplicated above, perhaps put it in User.

File: app/models/user.rb [hunk 2]
```diff
@@ -453,6 +454,16 @@ def set_default_widgets
     Dashboard::Manager.reset_user_to_default(self)
   end
 
+  def is_valid_route
```
Comment (domcleal) [inline app/models/user.rb:457]: `homepage_is_valid_route`

File: app/views/users/_form.html.erb [hunk 3]
```diff
@@ -28,6 +28,8 @@
 
       <%= time_zone_select_f(f, :timezone, Time.find_zone(@user.timezone) || Time.find_zone!('UTC'), { :include_blank => _('Browser timezone') }) %>
 
+      <%= text_f f, :homepage, placeholder: _('Leave blank for default'), help_inline: 'Redirect path after login' %>
```
Comment (domcleal) [inline app/views/users/_form.html.erb:31]: missing i18n

File: test/fixtures/settings.yml [hunk 4]
```diff
@@ -260,4 +260,10 @@ attributes55:
...
   default: 5
-  description: "Duration in minutes after the Puppet interval for servers to be classed as out of sync."
\ No newline at end of file
+  description: "Duration in minutes after the Puppet interval for servers to be classed as out of sync."
+attributes55:
+  name: home_page
+  settings_type: string
+  category: Setting::General
+  default: dashboard
+  description: "Path for home page that user lands on after login"
```
Comment (domcleal) [inline test/fixtures/settings.yml:269]: missing EOL

File: test/fixtures/settings.yml [hunk 5]
```diff
@@ -260,4 +260,10 @@ attributes55:
   settings_type: integer
   category: Setting::Puppet
   default: 5
-  description: "Duration in minutes after the Puppet interval for servers to be classed as out of sync."
\ No newline at end of file
+  description: "Duration in minutes after the Puppet interval for servers to be classed as out of sync."
+attributes55:
+  name: home_page
```
Comment (domcleal) [inline test/fixtures/settings.yml:265]: Hang on, this isn't used?

File: test/unit/setting_test.rb [hunk 6]
```diff
@@ -433,4 +433,18 @@ def check_value_returns_from_cache_with(options = {})
     Setting[name] = options[:value]
     assert_nil Rails.cache.read(name)
   end
+
+  test "should be valid to assign home_page to existing route path" do
```
Comment (domcleal) [inline test/unit/setting_test.rb:437]: This all looks out of date.

File: test/functional/home_controller_test.rb [hunk 7]
```diff
@@ -5,4 +5,22 @@ class HomeControllerTest < ActionController::TestCase
     get :status, {:format => "json"}
     assert_response :success
   end
+
+  test "should redirect to dashboard if home_page setting is blank" do
```
Comment (domcleal) [inline test/functional/home_controller_test.rb:9]: setting?  The user attribute?

File: test/functional/home_controller_test.rb [hunk 8]
```diff
@@ -5,4 +5,22 @@ class HomeControllerTest < ActionController::TestCase
...
+    get :index, {}, set_session_user
+    assert_redirected_to dashboard_url
+  end
+
+  test "should redirect to custom path if custom user homepage setting is set" do
+    assert User.find_by_login('secret_admin').update_attributes(:homepage => 'hostgroups')
```
Comment (domcleal) [inline test/functional/home_controller_test.rb:15]: Please don't modify the built in users, use FactoryGirl for tests.

**Comment by domcleal** [comment]
Tests are all red, please fix those.  A couple of other things:
1. Is there a better way to select a user's preferred login page apart from a text field?  Some sort of select_f perhaps?
2. Add homepage parameter to users API controllers.

**Comment by isratrade** [comment]
@domcleal, closing this for now. I may re-open later. There's a work-around within the fusor plugin.

**Pull request closed without merging**